File: cpp/test/test_csv.cpp
/*
 * test_csv.cpp - Test mpc_designation against CSV file of known conversions
 *
 * Usage: test_csv <csv_file> [max_errors]
 */

#include "../src/mpc_designation.hpp"
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <chrono>
#include <iomanip>

struct TestError {
    std::string input;
    std::string got;
    std::string expected;
};

bool runTests(const std::string& csvFile, int maxErrors) {
    std::ifstream fp(csvFile);
    if (!fp.is_open()) {
        std::cerr << "Error: Cannot open file: " << csvFile << "\n";
        return false;
    }

    long total = 0;
    long passed = 0;
    long failed = 0;
    std::vector<TestError> errors;

    auto startTime = std::chrono::high_resolution_clock::now();

    std::string line;
    // Skip header
    std::getline(fp, line);

    while (std::getline(fp, line)) {
        // Trim trailing whitespace
        while (!line.empty() && (line.back() == '\r' || line.back() == '\n')) {
            line.pop_back();
        }
        if (line.empty()) continue;

        // Skip header if encountered again
        if (line.substr(0, 8) == "unpacked") continue;

        total++;

        size_t comma = line.find(',');
        if (comma == std::string::npos) continue;

        std::string unpacked = line.substr(0, comma);
        std::string expectedPacked = line.substr(comma + 1);

        try {
            std::string gotPacked = mpc::MPCDesignation::convertSimple(unpacked);

            if (gotPacked != expectedPacked) {
                failed++;
                if (static_cast<int>(errors.size()) < maxErrors) {
                    errors.push_back({unpacked, gotPacked, expectedPacked});
                }
            } else {
                passed++;
            }
        } catch (const mpc::MPCDesignationError& e) {
            failed++;
            if (static_cast<int>(errors.size()) < maxErrors) {
                errors.push_back({unpacked, std::string("ERROR: ") + e.what(), expectedPacked});
            }
        }

        if (total % 100000 == 0) {
            std::cout << "Processed " << total << " entries...\n";
        }
    }

    auto endTime = std::chrono::high_resolution_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime).count();

    std::cout << "\n";
    std::cout << "=== Test Results ===\n";
    std::cout << "Total:  " << total << "\n";
    std::cout << "Passed: " << passed << "\n";
    std::cout << "Failed: " << failed << "\n";
    std::cout << "Time:   " << elapsed << "ms ("
              << std::fixed << std::setprecision(1) << (total * 1000.0 / elapsed)
              << " entries/sec)\n";
    std::cout << "\n";

    if (failed > 0) {
        std::cout << "=== First " << errors.size() << " failures ===\n";
        std::cout << std::left << std::setw(25) << "Input"
                  << std::setw(15) << "Got"
                  << std::setw(15) << "Expected" << "\n";
        std::cout << std::string(60, '-') << "\n";
        for (const auto& err : errors) {
            std::cout << std::left << std::setw(25) << err.input
                      << std::setw(15) << err.got
                      << std::setw(15) << err.expected << "\n";
        }
    }

    return failed == 0;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: test_csv <csv_file> [max_errors]\n";
        return 1;
    }

    std::string csvFile = argv[1];
    int maxErrors = (argc > 2) ? std::stoi(argv[2]) : 100;

    bool success = runTests(csvFile, maxErrors);
    return success ? 0 : 1;
}

File: csharp/test/TestCsv.cs
/*
 * TestCsv.cs - Test MPC designation against CSV file of known conversions
 *
 * Usage: dotnet run --project test/TestCsv.csproj <csv_file> [max_errors]
 */

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using MPC;

class TestCsv
{
    struct TestError
    {
        public string Input;
        public string Got;
        public string Expected;
    }

    static bool RunTests(string csvFile, int maxErrors)
    {
        if (!File.Exists(csvFile))
        {
            Console.Error.WriteLine($"Error: Cannot open file: {csvFile}");
            return false;
        }

        long total = 0;
        long passed = 0;
        long failed = 0;
        var errors = new List<TestError>();

        var sw = Stopwatch.StartNew();

        using (var reader = new StreamReader(csvFile))
        {
            // Skip header
            reader.ReadLine();

            string? line;
            while ((line = reader.ReadLine()) != null)
            {
                line = line.Trim();
                if (string.IsNullOrEmpty(line)) continue;

                // Skip header if encountered again
                if (line.StartsWith("unpacked")) continue;

                total++;

                int comma = line.IndexOf(',');
                if (comma < 0) continue;

                string unpacked = line.Substring(0, comma);
                string expectedPacked = line.Substring(comma + 1);

                try
                {
                    string gotPacked = MPCDesignation.ConvertSimple(unpacked);

                    if (gotPacked != expectedPacked)
                    {
                        failed++;
                        if (errors.Count < maxErrors)
                        {
                            errors.Add(new TestError { Input = unpacked, Got = gotPacked, Expected = expectedPacked });
                        }
                    }
                    else
                    {
                        passed++;
                    }
                }
                catch (MPCDesignationException e)
                {
                    failed++;
                    if (errors.Count < maxErrors)
                    {
                        errors.Add(new TestError { Input = unpacked, Got = $"ERROR: {e.Message}", Expected = expectedPacked });
                    }
                }

                if (total % 100000 == 0)
                {
                    Console.WriteLine($"Processed {total} entries...");
                }
            }
        }

        sw.Stop();
        long elapsed = sw.ElapsedMilliseconds;

        Console.WriteLine();
        Console.WriteLine("=== Test Results ===");
        Console.WriteLine($"Total:  {total}");
        Console.WriteLine($"Passed: {passed}");
        Console.WriteLine($"Failed: {failed}");
        Console.WriteLine($"Time:   {elapsed}ms ({total * 1000.0 / elapsed:F1} entries/sec)");
        Console.WriteLine();

        if (failed > 0)
        {
            Console.WriteLine($"=== First {errors.Count} failures ===");
            Console.WriteLine($"{"Input",-25} {"Got",-15} {"Expected",-15}");
            Console.WriteLine(new string('-', 60));
            foreach (var err in errors)
            {
                Console.WriteLine($"{err.Input,-25} {err.Got,-15} {err.Expected,-15}");
            }
        }

        return failed == 0;
    }

    static int Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.Error.WriteLine("Usage: TestCsv <csv_file> [max_errors]");
            return 1;
        }

        string csvFile = args[0];
        int maxErrors = args.Length > 1 ? int.Parse(args[1]) : 100;

        bool success = RunTests(csvFile, maxErrors);
        return success ? 0 : 1;
    }
}
